src/main/kotlin/org/jitsi/rtp/rtp/RedPacketParser.kt

```diff
@@ -0,0 +1,186 @@
+                            // packet. So we make the assumption that the redundant blocks represent the packets
+                            // directly proceeding the primary.
+                            sequenceNumber =
+                                (sequenceNumber - (redundancyBlockHeaders.size - index)).toRtpSequenceNumber().value
+                        }
+
+                packets?.add(redundancyPacket)
+            }
+            currentOffset += blockLength
+        }
+
+        // We've now read past the redundancy blocks, and currentOffset points to the payload of the primary block.
+        // Remove RED "in-place" by shifting the header.
+        val headerLength = headerLength
+        val newOffset = currentOffset - headerLength
+        val newLength = length - currentOffset + offset + headerLength
+        System.arraycopy(
+            buffer, offset,
+            buffer, newOffset,
+            headerLength
+        )
+        offset = newOffset
+        length = newLength
+        payloadType = primaryBlockHeader.pt.toInt()
+
+        return packets ?: emptyList()
+    }
+}
+
+sealed class BlockHeader(
+    /**
+     * The block's payload type.
+     */
+    val pt: Byte
+) {
+    abstract val headerLength: Int
+    companion object {
+        /**
+         * See RFC2198.
+         *
+         * Header blocks before the last (for redundant blocks):
+         *  0                   1                   2                   3
+         *  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+         * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+         * |1|   block PT  |  timestamp offset         |   block length    |
+         * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+         *
+         * The last header block (for the primary block):
+         *  0
+         *  0 1 2 3 4 5 6 7
+         * +-+-+-+-+-+-+-+-+
+         * |0|   block PT  |
+         * +-+-+-+-+-+-+-+-+
+         *
+         */
+        fun parse(buffer: ByteArray, offset: Int): BlockHeader {
+            val follow = buffer.getBitAsBool(offset, 0)
+            val pt = buffer.getBitsAsInt(offset, 1, 7).toByte()
+
+            if (!follow) return PrimaryBlockHeader(pt)
+
+            val timestampOffset =
+                (buffer[offset + 1].toPositiveInt() shl 6) + buffer.getBitsAsInt(offset + 2, 0, 6)
+            val blockLength =
+                (buffer.getBitsAsInt(offset + 2, 6, 2) shl 8) + buffer[offset + 3].toPositiveInt()
+
+            return RedundancyBlockHeader(pt, timestampOffset, blockLength)
+        }
+    }
+}
+
+class PrimaryBlockHeader(pt: Byte) : BlockHeader(pt) {
+    override val headerLength = 1
+}
+
+class RedundancyBlockHeader(
+    pt: Byte,
+    val timestampOffset: Int,
+    /**
+    * The length of the block's payload in bytes, or `null` if this is the last block (whose length is not explicitly
+    * specified).
```

Comment:
I think this comment is wrong?  The argument isn't nullable and this class doesn't apply to the last block.